src/olympia/versions/tasks.py

```diff
@@ -112,11 +113,95 @@ def delete_preview_files(pk, **kw):
+
+    for version in versions:
+        try:
+            log.info('Starting extraction of version "{}" for addon "{}" to '
+                     'git storage.'.format(version.pk, addon_pk))
+
+            extract_version_to_git(
+                version.pk,
+                # We do not want to trigger an infinite loop so an error will
+                # be thrown if there is a broken git reference.
+                stop_on_broken_ref=True,
+                # The `extract_version_to_git` task also checks the lock we set
+                # above so we need to tell it to skip this check here.
+                can_be_delayed=False,
+            )
+
+            log.info('Ending extraction of version "{}" for addon "{}" to '
+                     'git storage.'.format(version.pk, addon_pk))
+        except Exception:
+            log.exception('Error during extraction of version "{}" for addon '
+                          '"{}" to git storage.'.format(version.pk, addon_pk))
+            continue
+
+    # Remove the "git extraction lock" on this add-on.
+    git_storage.update(in_progress=False)
+    log.info(
+        'Ending extraction of addon "{}" to git storage.'.format(addon_pk)
+    )
+
+
+@task
+@use_primary_db
+def extract_version_to_git(version_id, author_id=None, note=None,
+                           stop_on_broken_ref=False, can_be_delayed=True):
+    """Extract a `File` into our git storage backend.
+
+    - `stop_on_broken_ref`: when we detect a broken git reference (when looking
+      up a git branch), the git repository is in a broken state and it is not
+      possible to recover from that state. In this case, we delete the git
+      repository, update all the versions and re-extract everything thanks to
+      the `extract_version_to_git` task, which uses `extract_version_to_git`
+      under the hood. This parameter is used to prevent a potential infinite
+      loop in case of another broken ref initiated by `extract_version_to_git`.
+
+    - `can_be_delayed`: each add-on object can be "locked" for git extraction
+      in `extract_addon_to_git()` so that we do not attempt to extract a
+      version while the add-on is being extracted. This parameter is used to
+      by-pass the lock check when `extract_addon_to_git()` calls this function.
+    """
     # We extract deleted or disabled versions as well so we need to make sure
     # we can access them.
     version = Version.unfiltered.get(pk=version_id)
+    addon = version.addon
+
+    if can_be_delayed and addon.git_extraction_is_in_progress:
+        log.info('Delaying task "extract_version_to_git" for version_id={} '
+                 'because the add-on is already being git-extracted.'
+                 ''.format(version_id))
+
+        extract_version_to_git.apply_async(
+            kwargs={
+                'version_id': version_id,
+                'author_id': author_id,
+                'note': note,
+                'stop_on_broken_ref': stop_on_broken_ref
```

Comment:
Just in case I'd propagate `can_be_delayed` as well (in case the default value ever changes)